mlir/lib/Dialect/Rock/IR/MfmaInsnGroup.cpp

```diff
@@ -395,24 +395,22 @@ FailureOr<MfmaInsnGroup> MfmaInsnGroup::select(mlir::Type elementType,
                           << "mPerWave: " << mPerWave << "\n"
                           << "nPerWave: " << nPerWave << "\n");
 
-  // Use 64x64 as base unit in large waves
-  int64_t mPerMfmaGroup = getLenPerMfmaGroup(mPerWave);
-  int64_t nPerMfmaGroup = getLenPerMfmaGroup(nPerWave);
+  // Decision on how we want to "compose" tiles: there is a maximal legnth,
+  // after which we want to iterate over each tile.
+  int64_t maximalLen = getMaximalLenPerWave(elementType);
+  int64_t mPerMfmaGroup = getLenPerMfmaGroup(mPerWave, maximalLen);
+  int64_t nPerMfmaGroup = getLenPerMfmaGroup(nPerWave, maximalLen);
 
   MfmaInsnGroupSelectKey key = {convertTypeToId(elementType), mPerMfmaGroup,
                                 nPerMfmaGroup};
   auto mfmaInsnGroupAttrMap = getMfmaInsnGroupAttrMap();
   auto it = mfmaInsnGroupAttrMap.find(key);
   if (it != mfmaInsnGroupAttrMap.end()) {
     MfmaInsnGroupAttr groupAttr = (*it).second;
-    // Override the repeat information in case this is for larger wave
-    if (mPerWave > 64) {
-      groupAttr.mRepeats = mPerWave / 64;
-    }
 
-    if (nPerWave > 64) {
-      groupAttr.nRepeats = nPerWave / 64;
-    }
+    // Override the repeat information in case this is for larger wave
+    groupAttr.mRepeats = mPerWave / maximalLen;
+    groupAttr.nRepeats = nPerWave / maximalLen;
```

Comment:
Instead of inventing more concept like `maximalLen` you can use an existing parameter and it would be able to generalize:

`groupAttr.mRepeats = mPerWave / (mfmaInsnAttr.mfmaNonKDim * mfmaInsnAttr.blocksMfma);`

Since 
 - the reduction ones' `blocksMfma` is always 1, it will be only a fragment of 64
 - the non-reduction ones' `blocksMfma * mfmaNonKDim` must be 64

This can make your repeats deduction logic to scale to all xdlops instructions.

Reply:
Yes, actually this was a bug. I can simply use `mPerMfmaGroup`. However, I still need the concept of `maximalLen`, because it will be used to pick the right `mfma`. Let me clarify. 

Suppose we have `MPerWave x NPerWave = <64,64>` . 

In our xdlops DB, we have:
1. <64,64> -> `mfma32x32x4f16` -> broadcast
2. <32,32> -> `mfma32x32x8f16` -> reduction

If `maximalLen=64`, this means we select entry `1`  - which is wrong. 

If I change the `<64,64>` entry in our DB to use a reduction and set `m/n repeats = <2,2>` I will cover a `<128x128>` tile size - still wrong. 

In this way, instead I have exactly what I want (and it is coherent with our infra). When I select `fp16` I want the maximal length to be `32x32` (which uses reduction) : anything bigger will be "composed" by `32x32` tiles.

Reply:
In my equation:
 - mfma32x32x4f16: mfmaNonKDim * blocksMfma = 64
 - mfma32x32x8f16: mfmaNonKDim * blocksMfma = 32

Reply:
Yes, I agree on the bug. So if you see the latest change it is:
```
    groupAttr.mRepeats = mPerWave / mPerMfmaGroup;
    groupAttr.nRepeats = nPerWave / nPerMfmaGroup;
```
I think that `mPerMfmaGroup==mfmaNonKDim * blocksMfma `, but anyway if `mPerWave < maximalLen` then `mPerMfmaGroup=mPerWave` and `groupAttr.mRepeats=1`. Otherwise, it computes the right `mRepeats`

Reply:
So, ok, after a chat I think we clarified everything. Basically `mPerWaves` and `nPerWaves` will act as keys for our database. Everything else needs to be derived from the `mfma` instruction we select. 

The only caveat is that for broadcasts, your formula could get `mRepeats` or `nRepeats` to zero (e.g., `32x64` or `64x32`). I simply did a `std::max(1, formula)` to cover those cases.